### PR TITLE
Improved Quarkus platform detection mechanism

### PR DESCRIPTION
**Pull request by aloubyansky**
Quarkus platform detection mechanism based on a "fingerprint" dependency constraint represented by the JSON platform descriptor artifact introduced into the platform BOM.

There are couple of reasons behind this PR:
1. The current way of detecting platforms used in Quarkus applications is inefficient and most importantly not very reliable (it is based on the assumption that we can detect all the BOM imports and check whether each imported BOM is a Quarkus platform);
2. Being able to identify imports of potentially conflicting platforms (conflicting on e.g. the Quarkus code versions they are based on) by the same application.

This PR changes the coordinates of the platform descriptor to `<platform-group-id>:<platform-artifact-id>-quarkus-platform-descriptor:<platform-version>:json:<platform-version>`.
Which, e.g. in case of the current `quarkus-bom`, it will be `io.quarkus:quarkus-bom-quarkus-platform-descriptor:999-SNAPSHOT:json:999-SNAPSHOT`.
The platform version in the middle of the example above is the artifact's classifier. The classifier is a part of the artifact's ID. Which means when the project's effective `dependencyManagement` section is resolved, all the platform imports will be preserved and in the proper order. The classifier in this case will allow to detect imports of different versions of the same platform in a given application (which is very much possible in case a project imports two platforms that are based on different versions of quarkus-bom).

This specific PR, however, does not introduce the platform conflict checks yet, although given that all the necessary data for this kind of check is already included and it'd be very easy to implement. For now it is only replacing the legacy platform detection mechanism making it more efficient and reliable.

### CODE REVIEW

File: independent-projects/tools/platform-descriptor-resolver-json/src/main/java/io/quarkus/platform/descriptor/resolver/json/QuarkusJsonPlatformDescriptorResolver.java [hunk 1]
```diff
@@ -58,21 +59,43 @@
...
+                        bomVersion, log)),
+                ((bomGroupId, bomArtifactId, bomVersion, jsonResolver, log) -> jsonResolver.resolveArtifact(bomGroupId,
+                        bomArtifactId, null, "json", bomVersion, log)),
+                ((bomGroupId, bomArtifactId, bomVersion, jsonResolver, log) -> jsonResolver.resolveArtifact(bomGroupId,
+                        bomArtifactId + "-descriptor-json", null, "json", bomVersion, log)));
+    }
+
+    public static JsonDescriptorResolver jsonDescriptorResolver() {
+        return jsonDescriptorResolver;
+    }
```
Comment (aloubyansky) [inline independent-projects/tools/platform-descriptor-resolver-json/src/main/java/io/quarkus/platform/descriptor/resolver/json/QuarkusJsonPlatformDescriptorResolver.java:76]: @gastaldi this PR changes the coordinates of the JSON descriptor and will require adjustments in the extension catalog project. To hopefully make it easier to maintain, I introduced this method and a couple interfaces right above this class. This specific method returns a JSON resolver that will try resolving the JSON for the BOM coordinates using the legacy and the new JSON coordinates conventions.

**Comment by aloubyansky** [comment]
Not sure who wants to review this. But this affects the tools. So for your info @quarkusio/devtools

**Comment by aloubyansky** [comment]
This is also touching the Gradle plugin, so also fyi @glefloch

**Review by glefloch — COMMENTED** [review]
I don't quite get all pieces involved by this change but the gradle part looks good to me.

File: independent-projects/tools/platform-descriptor-resolver-json/src/main/java/io/quarkus/platform/descriptor/resolver/json/QuarkusJsonPlatformDescriptorResolver.java (hunk 1 above)
Comment (gastaldi) [reply independent-projects/tools/platform-descriptor-resolver-json/src/main/java/io/quarkus/platform/descriptor/resolver/json/QuarkusJsonPlatformDescriptorResolver.java:76]: @aloubyansky No problem, it's just a matter of changing the `platforms.json` to the new GAV

File: devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusListExtensions.java [hunk 2]
```diff
@@ -96,7 +97,11 @@ public QuarkusListExtensions() {
     @TaskAction
     public void listExtensions() {
         try {
-            ListExtensions listExtensions = new ListExtensions(getQuarkusProject())
+            final QuarkusProject quarkusProject = getQuarkusProject();
+            getLogger().info("Quarkus platform " + quarkusProject.getPlatformDescriptor().getBomGroupId() + ":"
```
Comment (gastaldi) [inline devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusListExtensions.java:101]: Excellent! :clap:

File: devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusPlatformTask.java [hunk 3]
```diff
@@ -30,48 +39,148 @@
...
 
-        List<QuarkusPlatformDescriptor> platforms = new ArrayList<>(2);
+        final QuarkusPlatformDescriptor platform = resolvePlatformDescriptor(platformResolver, getProject(), boms);
+        if (platform != null) {
+            return platform;
+        }
+
+        final List<QuarkusPlatformDescriptor> platforms = resolveLegacyPlatforms(platformResolver, boms);
+        if (platforms.isEmpty()) {
+            throw new GradleException("Failed to determine Quarkus platform for the project");
```
Comment (gastaldi) [inline devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusPlatformTask.java:58]:
```suggestion
            throw new GradleException("Failed to determine the Quarkus platform for the project");
```

**Review by gastaldi — APPROVED** [review]
I think it looks good overall. I added a small suggestion

**Comment by gastaldi** [comment]
@aloubyansky is there any relation between this PR and https://github.com/quarkusio/quarkus/pull/12785?

**Comment by aloubyansky** [comment]
> @aloubyansky is there any relation between this PR and #12785?

At this point they are still loosely coupled. I.e. this change could help identify whether for a given found platform properties artifact there is a corresponding json descriptor and by that confirm that the properties belong to the platform imported by the project.

The original idea was to combine the properties and the json under the same GA but it can't be done for quarkus-bom because the properties have to be available early in the build to pass the tests while the JSON has to be generated after all the extensions have been built.

**Pull request merged**